patina_dxe_core/src/perf_timer.rs

```diff
@@ -0,0 +1,98 @@
+use patina::{component::service::IntoService, perf_timer::ArchTimerFunctionality};
+use spin::Once;
+
+/// Cached performance frequency.
+pub(crate) static PERF_FREQUENCY: Once<u64> = Once::new();
+
+/// Performance timer implementation.
+#[derive(IntoService)]
+#[service(dyn ArchTimerFunctionality)]
+pub struct PerfTimer;
+
+impl ArchTimerFunctionality for PerfTimer {
+    /// Value of the counter (ticks).
+    fn cpu_count(&self) -> u64 {
+        arch_cpu_count()
+    }
+
+    /// Frequency of `cpu_count` increments (in Hz).
+    /// If a platform has provided a custom frequency via `PERF_FREQUENCY`, that value is used.
+    /// Otherwise, an architecture-specific method is attempted to determine the frequency.
+    fn perf_frequency(&self) -> u64 {
+        if let Some(freq) = PERF_FREQUENCY.get() { *freq } else { arch_perf_frequency() }
+    }
+}
+
+/// Returns the current CPU count using architecture-specific methods.
+///
+/// Skip coverage as any value could be valid, including 0.
+#[coverage(off)]
+pub(crate) fn arch_cpu_count() -> u64 {
+    #[cfg(target_arch = "x86_64")]
+    {
+        use core::arch::x86_64;
+        unsafe { x86_64::_rdtsc() }
+    }
+    #[cfg(target_arch = "aarch64")]
+    {
+        use aarch64_cpu::registers::{self, Readable};
+        registers::CNTPCT_EL0.get()
+    }
+}
+
+/// Returns the performance frequency using architecture-specific methods.
+/// In general, the performance frequency is a configurable value that may be
+/// provided by the platform. This function is a fallback when no
+/// platform-specific configuration is provided.
+///
+/// Skip coverage as any value could be valid, including 0.
+#[coverage(off)]
+pub(crate) fn arch_perf_frequency() -> u64 {
+    let mut frequency = 0u64;
+
+    // Try to get TSC frequency from CPUID (most Intel and AMD platforms).
+    #[cfg(target_arch = "x86_64")]
+    {
+        use core::arch::x86_64;
+        use core::arch::x86_64::CpuidResult;
+
+        let CpuidResult { eax, ebx, ecx, .. } = unsafe { x86_64::__cpuid(0x15) };
+        if eax != 0 && ebx != 0 && ecx != 0 {
+            // CPUID 0x15 gives TSC_frequency = (ECX * EAX) / EBX.
+            // Most modern x86 platforms support this leaf.
+            frequency = (ecx as u64 * ebx as u64) / eax as u64;
+        } else {
+            // CPUID 0x16 gives base frequency in MHz in EAX.
+            // This is supported on some older x86 platforms.
+            // This is a nominal frequency and is less accurate for reflecting actual operating conditions.
+            let CpuidResult { eax, .. } = unsafe { x86_64::__cpuid(0x16) };
+            if eax != 0 {
+                frequency = (eax * 1_000_000) as u64;
+            }
+        }
+    }
+
+    // Use CNTFRQ_EL0 for aarch64 platforms.
+    #[cfg(target_arch = "aarch64")]
+    {
+        use aarch64_cpu::registers::{self, Readable};
+        frequency = registers::CNTFRQ_EL0.get();
```

Comment:
Consider using the new read_sysreg  in the SDK instead

https://github.com/OpenDevicePartnership/patina/blob/main/sdk/patina/src/arch/aarch64.rs